src/v/cluster/scheduling/leader_balancer_constraints.cc

```diff
@@ -10,15 +10,236 @@
+            }
+
+            for (const auto& g_info : node->second) {
+                const auto& leader = g_info.leader;
+                const auto& group = g_info.group_id;
+                const auto& replicas = g_info.replicas;
+
+                // Don't try moving any groups that are currently muted.
+                if (mi().muted_groups().contains(group)) {
+                    continue;
+                }
+
+                for (const auto& replica : replicas) {
+                    // Don't try a move to a different shard to on the same
+                    // node. As it won't decrease error
+                    if (replica.node_id == node->first || leader == replica) {
+                        continue;
+                    }
+
+                    // Don't try moving group to a muted node.
+                    if (mi().muted_nodes().contains(replica.node_id)) {
+                        continue;
+                    }
+
+                    reassignment r{group, leader, replica};
+
+                    if (evaluate_internal(r) > error_jitter) {
+                        return r;
+                    }
+                }
+            }
+        }
+    }
+
+    return std::nullopt;
+}
+
+void even_topic_distributon_constraint::rebuild_indexes() {
+    _topic_node_index.clear();
+    _topic_replica_index.clear();
+    _topic_partition_index.clear();
+
+    for (const auto& broker_shard : si().shards()) {
+        for (const auto& group_p : broker_shard.second) {
+            auto topic_id = group_to_topic_id().at(group_p.first);
+            const auto& node_id = broker_shard.first.node_id;
+
+            _topic_node_index[topic_id][node_id].emplace_back(
+              group_p.first, broker_shard.first, group_p.second);
+            _topic_partition_index[topic_id] += 1;
+
+            // Some of the replicas may not have leadership. So add
+            // all replica nodes here.
+            for (const auto& replica_bs : group_p.second) {
+                _topic_replica_index[topic_id].insert(replica_bs.node_id);
+                _topic_node_index[topic_id].try_emplace(replica_bs.node_id);
+            }
+        }
+    }
+}
+
+void even_topic_distributon_constraint::calc_topic_skew() {
+    _topic_skew.clear();
+    _topic_opt_leaders.clear();
+
+    for (const auto& topic : _topic_node_index) {
+        auto topic_partitions = static_cast<double>(
+          _topic_partition_index.at(topic.first));
+        auto topic_replicas = static_cast<double>(
+          _topic_replica_index.at(topic.first).size());
+        auto opt_leaders = topic_partitions
+                           / std::min(topic_replicas, topic_partitions);
+        _topic_opt_leaders[topic.first] = opt_leaders;
+
+        auto& skew = _topic_skew[topic.first];
+        skew = 0;
+
+        for (const auto& node : topic.second) {
+            auto leaders = static_cast<double>(node.second.size());
+
+            skew += pow(leaders - opt_leaders, 2);
```

Comment:
nit: I think the math here could use a comment in the header.

Reply:
Will add a comment to explain the equations used.